File: Web/Backend-API/FMS/Service/FarmService.py
# FarmService.py

from typing import List
from fastapi import Depends, HTTPException
from experiment.models import Farm, FarmIn, FarmProgress
from experiment.database import get_db


class FarmService:
    def __init__(self, db):
        self.db = db

    async def create_farm(self, farm_in: FarmIn) -> Farm:
        # This will check if the farm name already exists in the database
        farm = await self.db.farms.find_one({"name": farm_in.name})
        if farm:
            raise HTTPException(status_code=400, detail="Farm name already exists")

        # This will insert a new farm document into the database
        await self.db.farms.insert_one(farm_in.dict())

        return farm

    async def get_all_farms(self) -> List[Farm]:
        # This will query the database for all farms and return a list of Farm objects
        farms = await self.db.farms.find().to_list(length=100)
        return [Farm(**farm) for farm in farms]

    async def update_farm(self, farm_id: str, farm_in: FarmIn) -> Farm:
        # This will check if the farm exists in the database
        farm = await self.db.farms.find_one({"id": farm_id})
        if not farm:
            raise HTTPException(status_code=404, detail="Farm not found")

        # This will update the farm document in the database
        await self.db.farms.update_one({"id": farm_id}, {"$set": farm_in.dict()})

        return farm

    async def delete_farm(self, farm_id: str) -> None:
        # This will delete the farm from the database
        await self.db.farms.delete_one({"id": farm_id})

    async def get_farm_progress(self, farm_id: str) -> FarmProgress:
        # This will query the database for the farm progress and return a FarmProgress object
        farm_progress = await self.db.farm_progress.find_one({"farm_id": farm_id})
        if not farm_progress:
            raise HTTPException(status_code=404, detail="Farm progress not found")

        return farm_progress


farm_service = FarmService(get_db())

"""
This FarmService.py file implements the following methods:

create_farm(): This method creates a new farm in the database.
get_all_farms(): This method gets all farms from the database.
update_farm(): This method updates a farm in the database.
delete_farm(): This method deletes a farm from the database.
get_farm_progress(): This method gets the farm progress for a farm.
"""

File: Web/FMS/Services/FarmService.js
import api from "./Api"; // import your Api class

class FarmService {
  constructor() {
    // create an instance of the Api class
    this.api = api;
  }

  // define a method that creates a new farm
  async createFarm({ name, location, owner }) {
    try {
      // call the post method of the Api class with the url path and the farm data
      const data = await this.api.post(`/farms`, { name, location, owner });
      return data;
    } catch (error) {
      // handle the error and return null
      console.error(error);
      return null;
    }
  }

  // define a method that gets all farms
  async getAllFarms() {
    try {
      // call the get method of the Api class with the url path
      const data = await this.api.get(`/farms`);
      return data;
    } catch (error) {
      // handle the error and return null
      console.error(error);
      return null;
    }
  }

  // define a method that updates a farm
  async updateFarm({ id, name, location, owner }) {
    try {
      // call the put method of the Api class with the url path and the farm data
      const data = await this.api.put(`/farms/${id}`, {
        name,
        location,
        owner,
      });
      return data;
    } catch (error) {
      // handle the error and return null
      console.error(error);
      return null;
    }
  }

  // define a method that deletes a farm
  async deleteFarm({ id }) {
    try {
      // call the delete method of the Api class with the url path
      await this.api.delete(`/farms/${id}`);
    } catch (error) {
      // handle the error
      console.error(error);
    }
  }

  // define a method that gets a farm's progress
  async getFarmProgress({ id }) {
    try {
      // call the get method of the Api class with the url path
      const data = await this.api.get(`/farms/${id}/progress`);
      return data;
    } catch (error) {
      // handle the error and return null
      console.error(error);
      return null;
    }
  }
}

const farmService = new FarmService();

// export the instance as default or as a named export
export default farmService;

// export a new instance of the farmService class
//export default new farmService();

/*
This FarmService.js file provides the following methods:

createFarm(): Creates a new farm.
getAllFarms(): Gets all farms.
updateFarm(): Updates a farm.
deleteFarm(): Deletes a farm.
getFarmProgress(): Gets a farm's progress.
You can use these methods to implement farm management in your React.js project.
*/
